File: crossage_fr/match/age_gap.py
"""Cross-age gap estimation and NIST-grounded confidence banding.

The product is recall-first and human-in-the-loop because wide-gap child->adult
recognition is provably weak. NIST IFPC 2025 (Longitudinal Evaluation of Child Face
Recognition) reports TAR@0.1%FAR collapsing with the enrolment->probe age gap:
~98.5% at 2 years, ~95.7% at 4 years, ~87.2% at 6 years, ~71.3% at 8 years.

This module turns the gap between a reference photo's capture date and a candidate
photo's capture date into a coarse, honest confidence band. It NEVER asserts identity;
it only tells the reviewer how much weight the machine score deserves given the gap.
"""

from __future__ import annotations

from datetime import date, datetime

# (max_inclusive_years, confidence_band). Mirrors the NIST IFPC 2025 TAR cliff.
_BANDS: tuple[tuple[float, str], ...] = (
    (2.0, "high"),
    (4.0, "moderate"),
    (6.0, "low"),
)
_WIDE_GAP_BAND = "very-low"

# A gap at or beyond this many years earns the cross-age-gap review flag.
FLAG_THRESHOLD_YEARS = 4.0
CROSS_AGE_GAP_FLAG = "cross-age-gap"

_DATE_FORMATS = ("%Y-%m-%d", "%Y-%m-%dT%H:%M:%S", "%Y:%m:%d %H:%M:%S", "%Y-%m-%d %H:%M:%S")


def _parse_date(value: str | None) -> date | None:
    if not value or not isinstance(value, str):
        return None
    text = value.strip()
    if not text:
        return None
    # Tolerate a trailing Z / time component by trying progressively looser parses.
    candidate = text[:-1] if text.endswith("Z") else text
    for fmt in _DATE_FORMATS:
        try:
            return datetime.strptime(candidate, fmt).date()
        except ValueError:
            continue
    # Last resort: ISO date prefix (YYYY-MM-DD).
    try:
        return date.fromisoformat(candidate[:10])
    except ValueError:
        return None


def confidence_for_gap(years: float) -> str:
    """Return the NIST-grounded confidence band for an absolute age gap in years."""
    magnitude = abs(years)
    for ceiling, band in _BANDS:
        if magnitude <= ceiling:
            return band
    return _WIDE_GAP_BAND


ESTIMATED_BAND = "estimated"


def compute_age_gap(
    candidate_date: str | None,
    reference_date: str | None,
    *,
    candidate_provenance: str = "exif",
    reference_provenance: str = "exif",
) -> tuple[float | None, str | None, str | None]:
    """Compute (age_gap_years, confidence_band, review_flag).

    Returns (None, None, None) when either date is missing/unparseable — the feature is
    purely additive and must never block a candidate when dates are unknown.

    §5.4 governance: the NIST confidence band and the cross-age review flag are only
    trustworthy when BOTH dates are real EXIF event dates. If either date is the
    mtime fallback (the scan date for a digitized historical photo) the numeric gap
    may be meaningless, so we return the gap as informational with confidence
    "estimated" and NO review flag — the reviewer is told the gap is unverified
    rather than shown a false NIST reliability band.
    """
    cand = _parse_date(candidate_date)
    ref = _parse_date(reference_date)
    if cand is None or ref is None:
        return (None, None, None)
    years = round(abs((cand - ref).days) / 365.25, 2)
    verified = candidate_provenance == "exif" and reference_provenance == "exif"
    if not verified:
        return (years, ESTIMATED_BAND, None)
    confidence = confidence_for_gap(years)
    flag = CROSS_AGE_GAP_FLAG if years >= FLAG_THRESHOLD_YEARS else None
    return (years, confidence, flag)

File: mobile-app/packages/decision-layer/src/ageGap.ts
// Port of crossage_fr/match/age_gap.py — NIST-grounded cross-age confidence banding.
//
// This never asserts identity. It turns the gap between a reference photo's capture date and a
// candidate photo's capture date into an honest confidence band, and relaxes the REVIEW floor
// (never the confident/likely operating points) for verified wide gaps. This is our wedge:
// Apple Photos has no published cross-age mechanism.

// (maxInclusiveYears, band) — mirrors the NIST IFPC 2025 TAR cliff.
const BANDS: ReadonlyArray<readonly [number, string]> = [
  [2.0, 'high'],
  [4.0, 'moderate'],
  [6.0, 'low'],
];
const WIDE_GAP_BAND = 'very-low';
export const ESTIMATED_BAND = 'estimated';

export const FLAG_THRESHOLD_YEARS = 4.0;
export const CROSS_AGE_GAP_FLAG = 'cross-age-gap';
const MIN_CROSS_AGE_REVIEW_THRESHOLD = 0.12;

/**
 * Parse the date formats the reference tolerates: ISO, EXIF (`YYYY:MM:DD HH:MM:SS`), a trailing
 * Z, and a bare ISO date prefix. Returns a UTC-midnight epoch-day count, or null.
 */
export function parseDateToDays(value: string | null | undefined): number | null {
  if (!value || typeof value !== 'string') return null;
  const textRaw = value.trim();
  if (!textRaw) return null;
  const text = textRaw.endsWith('Z') ? textRaw.slice(0, -1) : textRaw;

  // EXIF `YYYY:MM:DD[ HH:MM:SS]` and ISO `YYYY-MM-DD[THH:MM:SS | space HH:MM:SS]`.
  const m = text.match(/^(\d{4})[:-](\d{2})[:-](\d{2})/);
  if (m) {
    const y = Number(m[1]);
    const mo = Number(m[2]);
    const d = Number(m[3]);
    // Validate real calendar date (reject 2020-13-40 etc.), matching strptime strictness.
    const utc = Date.UTC(y, mo - 1, d);
    const back = new Date(utc);
    if (back.getUTCFullYear() === y && back.getUTCMonth() === mo - 1 && back.getUTCDate() === d) {
      return Math.floor(utc / 86_400_000);
    }
  }
  return null;
}

/** NIST-grounded confidence band for an absolute age gap in years. */
export function confidenceForGap(years: number): string {
  const magnitude = Math.abs(years);
  for (const [ceiling, band] of BANDS) {
    if (magnitude <= ceiling) return band;
  }
  return WIDE_GAP_BAND;
}

export interface AgeGapResult {
  years: number | null;
  confidence: string | null;
  flag: string | null;
}

/**
 * Compute (ageGapYears, confidenceBand, reviewFlag). Returns all-null when either date is
 * missing/unparseable — the feature is purely additive and must never block a candidate.
 * The NIST band and cross-age flag are only emitted when BOTH dates are real EXIF event dates;
 * otherwise the gap is returned as informational ("estimated") with no flag.
 */
export function computeAgeGap(
  candidateDate: string | null | undefined,
  referenceDate: string | null | undefined,
  candidateProvenance = 'exif',
  referenceProvenance = 'exif',
): AgeGapResult {
  const cand = parseDateToDays(candidateDate);
  const ref = parseDateToDays(referenceDate);
  if (cand === null || ref === null) return { years: null, confidence: null, flag: null };
  // round(x, 2) — Python uses banker's rounding, but at 2 dp over day/365.25 the ties that
  // differ are astronomically unlikely; we match round-half-away and the fixtures confirm it.
  const years = roundTo(Math.abs(cand - ref) / 365.25, 2);
  const verified = candidateProvenance === 'exif' && referenceProvenance === 'exif';
  if (!verified) return { years, confidence: ESTIMATED_BAND, flag: null };
  const confidence = confidenceForGap(years);
  const flag = years >= FLAG_THRESHOLD_YEARS ? CROSS_AGE_GAP_FLAG : null;
  return { years, confidence, flag };
}

/** Review-only threshold floor for a verified wide gap. Strong/likely operating points never move. */
export function reviewThresholdForGap(
  baseThreshold: number,
  years: number | null | undefined,
  confidence: string | null | undefined,
): number {
  const base = Number(baseThreshold);
  if (!Number.isFinite(base)) return baseThreshold;
  const gap = years == null ? 0.0 : Math.abs(Number(years));
  if (confidence === 'very-low' && gap > 6.0) {
    return roundTo(Math.max(MIN_CROSS_AGE_REVIEW_THRESHOLD, base - 0.04), 6);
  }
  if (confidence === 'low' && gap > 4.0) {
    return roundTo(Math.max(MIN_CROSS_AGE_REVIEW_THRESHOLD, base - 0.02), 6);
  }
  return roundTo(base, 6);
}

/** Round-half-away-from-zero to `digits` decimals, matching Python's round() for these fixtures. */
function roundTo(value: number, digits: number): number {
  const factor = 10 ** digits;
  return Math.round((value + Number.EPSILON) * factor) / factor;
}
